drivers/net/wireless/marvell/mwifiex/pcie_quirks.h

```diff
@@ -0,0 +1,104 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Header file for PCIe quirks.
+ */
+
+#include <linux/dmi.h>
+
+/* List of Surface gen4+ (Pro 4 or later gen) devices that have mwifiex */
+static const struct dmi_system_id surface_gen4p_dmi_table[] = {
+	{
+		.ident = "Surface Pro 4",
+		.matches = {
+			DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"),
+			DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Surface Pro 4"),
+		},
+	},
+	{
+		.ident = "Surface Pro 5",
+		.matches = {
+			/* match for SKU here due to generic product name "Surface Pro" */
+			DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"),
+			DMI_EXACT_MATCH(DMI_PRODUCT_SKU, "Surface_Pro_1796"),
+		},
+	},
+	{
+		.ident = "Surface Pro 5 (LTE)",
+		.matches = {
+			/* match for SKU here due to generic product name "Surface Pro" */
+			DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"),
+			DMI_EXACT_MATCH(DMI_PRODUCT_SKU, "Surface_Pro_1807"),
+		},
+	},
+	{
+		.ident = "Surface Pro 6",
+		.matches = {
+			DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"),
+			DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Surface Pro 6"),
+		},
+	},
+	{
+		.ident = "Surface Book 1",
+		.matches = {
+			DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"),
+			DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Surface Book"),
+		},
+	},
+	{
+		.ident = "Surface Book 2",
+		.matches = {
+			DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"),
+			DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Surface Book 2"),
+		},
+	},
+	{
+		.ident = "Surface Laptop 1",
+		.matches = {
+			DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"),
+			DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Surface Laptop"),
+		},
+	},
+	{
+		.ident = "Surface Laptop 2",
+		.matches = {
+			DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"),
+			DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Surface Laptop 2"),
+		},
+	},
+	{}
+};
+
+/* Surface 3 */
+static const struct dmi_system_id surface3_dmi_table[] = {
+	{
+		.ident = "Surface 3",
+		.matches = {
+			DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"),
+			DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Surface 3"),
+		},
+	},
+	{
+		.ident = "Surface 3",
+		.matches = {
+			DMI_EXACT_MATCH(DMI_BIOS_VENDOR, "American Megatrends Inc."),
+			DMI_EXACT_MATCH(DMI_SYS_VENDOR, "OEMB"),
+			DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "OEMB"),
+		},
+	},
+	{}
+};
+
+/* Surface Pro 3 */
+static const struct dmi_system_id surfacepro3_dmi_table[] = {
+	{
+		.ident = "Surface Pro 3",
+		.matches = {
+			DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"),
+			DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Surface Pro 3"),
+		},
+	},
+	{}
+};
```

Comment:
With bitflags you could combine all of those entries into one table, then match them at the beginning and set the bitflags of the card correspondingly. You can use the `driver_data` member to store a reference to the quirks.

Also it's generally a bad idea to have global `static const` variables in your header file. This means that they will be included in every compilation unit, so they will be duplicated. The usual way to set up global variables is via, in your `.h` file
```c
extern const struct dmi_system_id surface3_dmi_table[];
```
and in one of your `.c` files:
```c
const struct dmi_system_id surface3_dmi_table[] = {
    // ...
};
```

In this case though, it's probably better if you make the quirks fully private to the `.c` file and add a function like `mwifiex_initialize_quirks()` or something to set up the bitflags on `card`.

Reply:
> With bitflags you could combine all of those entries into one table, then match them at the beginning and set the bitflags of the card correspondingly. You can use the `driver_data` member to store a reference to the quirks.

OK, I've switched to using `driver_data` and combined all the entries into one table.

> Also it's generally a bad idea to have global `static const` variables in your header file. This means that they will be included in every compilation unit, so they will be duplicated. The usual way to set up global variables is via, in your `.h` file
> 
> ```c
> extern const struct dmi_system_id surface3_dmi_table[];
> ```
> 
> and in one of your `.c` files:
> 
> ```c
> const struct dmi_system_id surface3_dmi_table[] = {
>     // ...
> };
> ```

Thanks. For record, indeed I see only one place in upstream tree where DMI table exists in `.h` file (https://github.com/torvalds/linux/blob/master/drivers/input/serio/i8042-x86ia64io.h). It uses `__initconst`.

> In this case though, it's probably better if you make the quirks fully private to the `.c` file and add a function like `mwifiex_initialize_quirks()` or something to set up the bitflags on `card`.

OK, I've placed DMI table into `pcie_quirks.c` instead. The quirks will be set up by `mwifiex_initialize_quirks()`

Addressed by commit https://github.com/linux-surface/kernel/pull/58/commits/206fa3cd1d885b4bb5d72b8b518a2b90d580d548